### DISCUSSION TITLE
MAC address as byte string problems

### DISCUSSION THREAD

**Discussion by doceave** [micropython/micropython#11962] [Libraries & Drivers]
I am attempting to implement an ESPnow as described by this page: https://docs.petoi.com/apis/8266-micropython-controller/using-the-esp-now-protocol

I am using latest ESPnow firmware for an ESP32-S2

When I update the example to reflect the MAC address of my devices and run the code in terminal everything works. If, however, I upload the code to the device and allow it to run as main.py an error results.

It appears that the following line as at fault:

peer = b'\x48\x27\xe2\x46\xfe\xc4' 

When this line is uploaded it appears to be saved as:

peer1 = (b'H'âFþ¸')

Why is this particular line not being saved verbatim as is the rest of the file?

**Comment by robert-hh** [comment]
> When this line is uploaded it

How do you upload the file?

**Reply by doceave** [reply]
I have attempted two different methods which yield the same results:
> Ampy
> Arduino Lab for Micropython (which probably uses the above)

**Reply by bixb922** [reply]
Please try with mpremote. I haven't seen this kind of problems with that tool. What editor/IDE are you using?

**Reply by glenn20** [reply]
I endorse @bixb922 suggestion to use mpremote for coping files to your device. However, an alternative is to use:
```py
peer = bytes.fromhex("4827e246fec4")
```
rather than messing about with hex encoded byte strings.

Similarly, you can use `print(peer.hex())` (or `print(peer.hex("_"))`) to get a cleaner printout of the mac address. No need for `hexlify` since these methods were added to the `bytes` class.

**Reply by glenn20** [reply]
Whoops - should be `bytes.fromhex()` not `bytes.from_hex()` - updated comment above to fix.

**Comment by doceave** [comment]
My apologies for not making reply sooner --- I found two solutions which both point to Ampy and byte string handling as the issue:
> Using the Thonny IDE and the built in file transfer system --- no issues
> Using rshell --- no issues

Thanks for you assistance!
